test/integration/tcpingress_test.go

```diff
@@ -23,8 +24,19 @@ import (
 	"github.com/kong/kubernetes-ingress-controller/pkg/clientset"
 )
 
+var tcpMutex sync.Mutex
+
 func TestTCPIngressEssentials(t *testing.T) {
 	t.Parallel()
+	// Ensure no other TCP tests run concurrently to avoid fights over the port
+	// Free it when done
+	t.Log("locking TCP port")
+	tcpMutex.Lock()
+	defer func() {
+		t.Log("unlocking TCP port")
+		tcpMutex.Unlock()
+	}()
+
```

Comment:
In general :+1:, but I think now may be a good a time as any for us to simply fix the static ports issue in the tests such that there are multiple TCP ports available to test with. Let me know your thoughts? I would be happy to pair on this.

Reply:
My thinking was that expanding the available ports would mean that we'd then be adding a unique port for every test (making KTF overly specific to the specific needs of the KIC tests) or some sort of complex pool mechanism to obtain a port and lock it, or wait until one became available (kinda still the locking mechanism proposed here, but going from "up to 1 TCP test running at once" to "up to N TCP tests running at once").

Given the low impact to parallelism we'd have with this change (at most we block 2 tests for a while if the TCP+UDP test runs first), the latter option didn't seem worth the effort.

Reply:
I think we could just as easily add a pool of say, 3 available ports (for the 3 different tests here) and a locking mechanism so that tests can claim them from the environment. This could be added onto the Kong `Addon` implementation in the KTF, such that you can say something like this in the test suite:

```go
kongaddon := kong.NewBuilder()
    .WithTCPPortPool(3)
    .WithUDPPortPool(3)
    .Build()
```

And the tests could reference the ports like this:

```go
tcpPort, err := kongAddon.GetTCPPort()
udpPort, err := kongAddon.GetUDPPort()
```

The underlying implementation could basically consider those resourced locked until released:

```go
tcpPort, err := kongAddon.GetTCPPort()
defer kongAddon.ReleaseTCPPort(tcpPort)
```

This is just a mockup, but the basic idea being the `WithTCPPorts(total int)` function would just trigger additional ENV vars and flags for the Kong chart deployment so when the `environments.Environment` deploys the addon the cluster it has the N ports added, and `GetTCPPort()` and `ReleaseTCPPort()` would just be a simple threadsafe locking mechanism around these ports kep internally in a list within the `kongAddon` object.

LMKWYT?

Reply:
KIC is deployed on a customer cluster, would be hard to allocate TCP/UDP port arrange ahead. 
Pick a port from the reserved range 0 - 1023, check for conflicts would be the feasible solution in the tests. Currently, we are assuming no Test Ports conflict, which could not be true.

Reply:
> KIC is deployed on a customer cluster, would be hard to allocate TCP/UDP port arrange ahead.

True, but this isn't relevant as we're specifically talking about doing this for the testing environment.

> Pick a port from the reserved range 0 - 1023, check for conflicts would be the feasible solution in the tests.
> Currently, we are assuming no Test Ports conflict, which could not be true.

A default range of ports is technically a feasible alternative, albeit in many ways this is what we already kinda have but with a set of 1 port.

I believe that having the addon object responsible for providing access to semi-arbitrary ports is more ideal as it would allow the specifics of the port setup to be an implementation detail, and the callers don't really need _specific_ ports to test TCP/UDP.

Reply:
well, what I said is slightly different : even in test environment, port conflict could not resolve by addon object - you are not allowed to use 8000, 8888, 9000, 9999, which could be well-known kong service port. And we have more such service port.

Reply:
If we can't spare the time right now: If we at least create a detailed issue in the KTF that references this PR and explains that we need the ability to have multiple and dynamic ports available for testing TCP/UDP, I would consider that resolution for this comment for now so we can move forward (and use the lock for now).

Reply:
https://github.com/Kong/kubernetes-testing-framework/issues/93